# Плотные векторные представления слов

## Word2Vec

In [ ]:
import gensim.downloader

### Предварительно обученные векторы Word2Vec для английского языка

In [ ]:
word2vec_eng = gensim.downloader.load('word2vec-google-news-300')

[==================================================] 100.0% 1662.8/1662.8MB downloaded


In [ ]:
word2vec_eng['test']

array([-1.42578125e-01, -3.68652344e-02,  1.35742188e-01, -6.20117188e-02,
        7.95898438e-02,  1.90429688e-02, -8.15429688e-02, -1.27929688e-01,
       -2.95410156e-02,  2.36328125e-01, -1.21582031e-01, -2.14843750e-01,
        1.29882812e-01, -2.70996094e-02, -5.20019531e-02,  2.15820312e-01,
       -1.81640625e-01,  5.10253906e-02, -1.60156250e-01, -1.76757812e-01,
        1.83105469e-02, -4.12597656e-02, -2.32421875e-01, -1.03149414e-02,
        1.45507812e-01,  5.24902344e-02, -3.96484375e-01, -1.92871094e-02,
        2.51770020e-03, -1.26953125e-02, -4.39453125e-02,  3.07617188e-02,
        9.57031250e-02, -1.75781250e-01,  1.04370117e-02,  1.89453125e-01,
       -2.36328125e-01,  4.37011719e-02,  2.81250000e-01, -2.07519531e-02,
       -1.81640625e-01, -2.17773438e-01,  2.33398438e-01,  5.29785156e-02,
       -1.13769531e-01,  9.39941406e-03, -1.49414062e-01,  1.99218750e-01,
       -1.75781250e-01,  3.16406250e-01,  8.10546875e-02, -6.12792969e-02,
       -1.52343750e-01, -

In [ ]:
word2vec_eng['king']

array([ 1.25976562e-01,  2.97851562e-02,  8.60595703e-03,  1.39648438e-01,
       -2.56347656e-02, -3.61328125e-02,  1.11816406e-01, -1.98242188e-01,
        5.12695312e-02,  3.63281250e-01, -2.42187500e-01, -3.02734375e-01,
       -1.77734375e-01, -2.49023438e-02, -1.67968750e-01, -1.69921875e-01,
        3.46679688e-02,  5.21850586e-03,  4.63867188e-02,  1.28906250e-01,
        1.36718750e-01,  1.12792969e-01,  5.95703125e-02,  1.36718750e-01,
        1.01074219e-01, -1.76757812e-01, -2.51953125e-01,  5.98144531e-02,
        3.41796875e-01, -3.11279297e-02,  1.04492188e-01,  6.17675781e-02,
        1.24511719e-01,  4.00390625e-01, -3.22265625e-01,  8.39843750e-02,
        3.90625000e-02,  5.85937500e-03,  7.03125000e-02,  1.72851562e-01,
        1.38671875e-01, -2.31445312e-01,  2.83203125e-01,  1.42578125e-01,
        3.41796875e-01, -2.39257812e-02, -1.09863281e-01,  3.32031250e-02,
       -5.46875000e-02,  1.53198242e-02, -1.62109375e-01,  1.58203125e-01,
       -2.59765625e-01,  

In [ ]:
del word2vec_eng

NameError: ignored

### Предварительно обученные векторы Word2Vec для русского языка

Доступные предварительно обученные плотные векторные представления в библиотеки Gensim

In [ ]:
list(gensim.downloader.info()['models'].keys())

['fasttext-wiki-news-subwords-300',
 'conceptnet-numberbatch-17-06-300',
 'word2vec-ruscorpora-300',
 'word2vec-google-news-300',
 'glove-wiki-gigaword-50',
 'glove-wiki-gigaword-100',
 'glove-wiki-gigaword-200',
 'glove-wiki-gigaword-300',
 'glove-twitter-25',
 'glove-twitter-50',
 'glove-twitter-100',
 'glove-twitter-200',
 '__testing_word2vec-matrix-synopsis']

In [ ]:
word2vec_rus = gensim.downloader.load('word2vec-ruscorpora-300')

[==================================================] 100.0% 198.8/198.8MB downloaded


In [ ]:
word2vec_rus['тест']

KeyError: ignored

In [ ]:
for index, word in enumerate(word2vec_rus.index_to_key ):
    if index == 20:
        break
    print(f"word #{index}/{len(word2vec_rus.index_to_key )} is {word}")

word #0/184973 is весь_DET
word #1/184973 is человек_NOUN
word #2/184973 is мочь_VERB
word #3/184973 is год_NOUN
word #4/184973 is сказать_VERB
word #5/184973 is время_NOUN
word #6/184973 is говорить_VERB
word #7/184973 is становиться_VERB
word #8/184973 is знать_VERB
word #9/184973 is самый_DET
word #10/184973 is дело_NOUN
word #11/184973 is день_NOUN
word #12/184973 is жизнь_NOUN
word #13/184973 is рука_NOUN
word #14/184973 is очень_ADV
word #15/184973 is первый_ADJ
word #16/184973 is давать_VERB
word #17/184973 is новый_ADJ
word #18/184973 is слово_NOUN
word #19/184973 is иметь_VERB


In [ ]:
word2vec_rus['печь_VERB']

array([ 2.18535755e-02, -1.36514874e-02,  6.48699515e-03,  4.94286232e-02,
        4.72369529e-02,  2.99715698e-02,  1.84931476e-02,  5.63697964e-02,
        9.00735930e-02,  4.98076454e-02, -1.42216729e-02, -1.19277993e-02,
        4.64369878e-02, -2.20510038e-03, -5.23579530e-02,  6.86843460e-03,
        2.47351658e-02, -2.94565707e-02, -3.43381912e-02, -6.68611825e-02,
        2.16565677e-03, -3.44541185e-02,  5.10950834e-02,  3.39556672e-02,
       -1.03159972e-01,  1.10369720e-01, -4.85999370e-03, -9.61249918e-02,
       -4.24017571e-02, -6.25817403e-02,  7.29365274e-02, -3.05048935e-02,
       -1.22691244e-01,  5.59408627e-02, -3.77330072e-02, -5.60487658e-02,
        2.35838126e-02, -9.34875384e-03, -1.72916036e-02, -1.21203892e-01,
        5.48743568e-02,  4.17090692e-02,  9.80748981e-02, -1.06346466e-01,
       -2.49845888e-02, -9.09208506e-02,  7.08984062e-02,  2.82790884e-02,
        5.33705652e-02,  1.25205163e-02,  1.88979451e-02,  1.18532305e-04,
        5.06328084e-02,  

In [ ]:
word2vec_rus['печь_NOUN']

array([ 9.04111098e-03,  3.17178331e-02,  5.26587218e-02, -2.81463489e-02,
        3.36806215e-02, -1.19832382e-02,  6.69362992e-02,  5.34342714e-02,
        7.75137171e-02,  5.08050360e-02,  3.30233364e-03, -2.16485690e-02,
        5.27435094e-02,  1.38199255e-02, -7.34620243e-02, -2.33334582e-02,
        3.33712399e-02,  3.48375104e-02,  3.63502614e-02, -4.44506891e-02,
        1.17149157e-02, -4.02530469e-02, -6.20140284e-02,  7.47649819e-02,
       -3.94869968e-02, -4.06278297e-02, -8.34002271e-02, -6.80526569e-02,
       -1.62618563e-01, -5.65643683e-02,  1.65630854e-03, -1.48861082e-02,
       -3.62066291e-02,  3.73510309e-02, -2.62216832e-02, -3.96667421e-02,
       -9.02666152e-03,  1.10610299e-01, -1.12228990e-01, -1.61341488e-01,
        1.11707663e-02, -6.39451202e-03,  5.29266559e-02, -1.17891423e-01,
       -8.77382420e-03, -2.01879211e-05,  9.10191685e-02,  2.39265133e-02,
        7.54997283e-02, -1.52666410e-02,  5.74746467e-02, -5.05433492e-02,
        2.91562006e-02, -

In [ ]:
word2vec_rus.most_similar(['печь_NOUN'])

[('печка_NOUN', 0.8358493447303772),
 ('подтопок_NOUN', 0.7017349600791931),
 ('печной_ADJ', 0.6326410174369812),
 ('полати_NOUN', 0.6188919544219971),
 ('топиться_VERB', 0.6180375814437866),
 ('горнушка_NOUN', 0.6172196865081787),
 ('топиться::по-черному_VERB', 0.6171698570251465),
 ('истапливать_VERB', 0.6131374835968018),
 ('топка_NOUN', 0.6026756763458252),
 ('печурка_NOUN', 0.6010895371437073)]

In [ ]:
word2vec_rus.most_similar(['печь_VERB'])

[('напекать_VERB', 0.6473735570907593),
 ('испечь_VERB', 0.6360713243484497),
 ('жарить_VERB', 0.629960834980011),
 ('нардек_NOUN', 0.5905025005340576),
 ('выпекать_VERB', 0.5834865570068359),
 ('яшный_ADJ', 0.5677016377449036),
 ('варить_VERB', 0.5661610960960388),
 ('пирог_NOUN', 0.5652977228164673),
 ('тандыр_NOUN', 0.5573609471321106),
 ('месить::тесто_VERB', 0.5557463765144348)]

In [ ]:
del word2vec_rus

## fastText

Установка библиотеки fastText

In [ ]:
!git clone https://github.com/facebookresearch/fastText.git

Cloning into 'fastText'...
remote: Enumerating objects: 3953, done.
remote: Counting objects: 100% (1012/1012), done.
remote: Compressing objects: 100% (157/157), done.
remote: Total 3953 (delta 907), reused 867 (delta 853), pack-reused 2941
Receiving objects: 100% (3953/3953), 8.26 MiB | 33.69 MiB/s, done.
Resolving deltas: 100% (2514/2514), done.


In [ ]:
!cd fastText; pip install .

Processing /content/fastText
  Preparing metadata (setup.py) ... done
  Using cached pybind11-2.11.1-py3-none-any.whl (227 kB)
  Created wheel for fasttext: filename=fasttext-0.9.2-cp310-cp310-linux_x86_64.whl size=4291883 sha256=4cea4a1d6d6f9ad6e934d110313985eb49050accd48e147d029a580cf43a8766
  Stored in directory: /tmp/pip-ephem-wheel-cache-4gaarikk/wheels/8b/05/af/3cfae069d904597d44b309c956601b611bdf8967bcbe968903
Successfully built fasttext


In [ ]:
import fasttext.util

Загрузка предварительно обученных векторных представлений fastText для русского языка

In [ ]:
fasttext.util.download_model('ru', if_exists='ignore')

'cc.ru.300.bin'

In [ ]:
ft = fasttext.load_model('cc.ru.300.bin')

Показываем векторы для слов

In [ ]:
ft.get_word_vector('печь')

array([-1.58141226e-01,  1.40892416e-01, -1.16081268e-01, -3.65050286e-02,
        1.24294966e-01,  3.56620317e-03,  1.00102514e-01, -5.51618822e-03,
       -1.54162765e-01, -2.05441654e-01, -1.23218819e-01,  7.57891387e-02,
        3.42986137e-02, -4.14514728e-02, -4.02112976e-02,  2.98824906e-02,
       -5.66644147e-02,  8.71236175e-02,  4.90642563e-02,  1.29385591e-01,
        1.04177101e-02, -1.39765948e-01,  1.96572430e-02,  9.35998783e-02,
        2.53893584e-02, -8.46895576e-02, -2.33695209e-01, -9.06276107e-02,
       -3.56095424e-03,  1.33594170e-01,  1.00002609e-01, -1.07181132e-01,
        4.29465696e-02,  5.93492389e-03, -7.90850967e-02,  3.64862420e-02,
        1.04223136e-02, -1.24680214e-01, -1.95677444e-01, -5.12315333e-02,
        1.05863772e-01, -2.60896739e-02,  1.22163713e-01,  5.37529886e-02,
        4.78603020e-02,  1.70780927e-01,  2.17139110e-01, -1.11144893e-01,
        8.68928060e-02,  1.47899717e-01,  1.56384453e-01,  1.39447808e-01,
       -4.18573767e-02,  

In [ ]:
ft.get_word_vector('печи')

array([-1.70682058e-01,  1.00130618e-01, -6.11777566e-02, -3.40710282e-02,
        1.30481809e-01, -1.05290171e-02,  1.75007850e-01, -5.02382144e-02,
       -6.25204369e-02, -1.94488436e-01, -1.08190805e-01,  3.25171016e-02,
       -6.57913238e-02, -1.19659007e-01, -5.13470769e-02,  5.12411371e-02,
       -4.65683639e-02,  1.64096445e-01,  1.15095638e-01,  1.64232433e-01,
        2.63091996e-02, -4.70009372e-02,  3.41429412e-02,  3.52392346e-02,
        2.61768550e-02,  2.73864102e-02, -9.09000188e-02, -1.66930541e-01,
       -3.61857325e-04,  1.27198294e-01,  1.03386000e-01, -9.84837934e-02,
        4.89381216e-02, -4.85854596e-02, -6.08835034e-02,  6.37387261e-02,
        5.90716228e-02, -1.10543340e-01, -8.68958682e-02, -1.73928272e-02,
        1.08276293e-01, -2.56495290e-02,  1.41575366e-01, -3.85857895e-02,
        2.53301095e-02,  1.96836263e-01,  6.61053509e-02, -1.48278205e-02,
        5.18025607e-02,  6.68320283e-02,  8.72187465e-02,  9.31333154e-02,
       -5.87206185e-02, -

In [ ]:
ft.get_nearest_neighbors('печь')

[(0.7586913704872131, 'печь.'),
 (0.7371419668197632, 'печи'),
 (0.7335948944091797, 'печь-'),
 (0.7106799483299255, 'электропечь'),
 (0.6893130540847778, 'печку'),
 (0.6534967422485352, 'печка'),
 (0.6468836069107056, 'печью'),
 (0.6424350142478943, 'печь-каменка'),
 (0.638362467288971, 'Печь'),
 (0.631331741809845, 'буржуйка')]

## Navec

Библиотека [Navec](https://github.com/natasha/navec) — часть [проекта Natasha](https://github.com/natasha), коллекция предобученных плотных векторных представлений слов для русского языка. Качество решения сравнимо или выше, чем у статических моделей от RusVectores, размер в 5–6 раз меньше [6].

In [ ]:
!pip install navec

Загружаем файл с плотными векторным представлением слов

In [ ]:
!wget https://storage.yandexcloud.net/natasha-navec/packs/navec_hudlit_v1_12B_500K_300d_100q.tar -O navec_hudlit_v1_12B_500K_300d_100q.tar

--2023-09-30 07:34:52--  https://storage.yandexcloud.net/natasha-navec/packs/navec_hudlit_v1_12B_500K_300d_100q.tar
Resolving storage.yandexcloud.net (storage.yandexcloud.net)... 213.180.193.243, 2a02:6b8::1d9
Connecting to storage.yandexcloud.net (storage.yandexcloud.net)|213.180.193.243|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 53012480 (51M) [application/x-tar]
Saving to: ‘navec_hudlit_v1_12B_500K_300d_100q.tar’

navec_hudlit_v1_12B 100%[===================>]  50.56M  21.8MB/s    in 2.3s    

2023-09-30 07:34:55 (21.8 MB/s) - ‘navec_hudlit_v1_12B_500K_300d_100q.tar’ saved [53012480/53012480]



In [ ]:
from navec import Navec

In [ ]:
navec = Navec.load('navec_hudlit_v1_12B_500K_300d_100q.tar')

In [ ]:
navec['тест']

array([ 3.40839595e-01, -3.09203577e-04, -4.69992280e-01,  7.30743781e-02,
       -5.66504896e-01, -4.82852161e-01,  7.25645423e-01,  4.65745449e-01,
        7.27563053e-02, -2.22528860e-01, -4.74203676e-01,  1.69839531e-01,
        2.57203728e-01,  1.23394869e-01, -1.46565095e-01,  3.14269453e-01,
        1.57288715e-01, -8.49631310e-01, -7.16777444e-01,  5.74199498e-01,
       -2.25155413e-01,  5.71982004e-02,  3.24196637e-01,  4.09328550e-01,
        1.08715119e-02,  4.19131294e-02, -6.69988930e-01,  5.22277117e-01,
        8.90961066e-02, -4.72322136e-01,  2.04335764e-01,  4.24774468e-01,
        3.37487847e-01,  3.87795836e-01, -2.89813995e-01, -7.55393803e-02,
        2.62102157e-01,  4.13020432e-01,  3.93356323e-01, -5.41508645e-02,
        3.75927627e-01,  1.93368480e-01, -2.74716347e-01, -1.26420036e-01,
        5.58198154e-01, -6.14145637e-01,  2.10022889e-02,  1.14859723e-01,
       -1.69688568e-01,  2.09147960e-01, -3.22890967e-01, -2.60310590e-01,
        1.67034850e-01,  

### Плотное векторное представление Navec для отзывов на банки

In [ ]:
!pip install pymorphy3

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.4/55.4 kB 956.8 kB/s eta 0:00:00
  Using cached DAWG_Python-0.7.2-py2.py3-none-any.whl (11 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.4/8.4 MB 29.5 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import pymorphy3
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

In [ ]:
nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [ ]:
def preprocess(text, stop_words, punctuation_marks, morph):
    tokens = word_tokenize(text.lower())
    preprocessed_text = []
    for token in tokens:
        if token not in punctuation_marks:
            lemma = morph.parse(token)[0].normal_form
            if lemma not in stop_words:
                preprocessed_text.append(lemma)
    return preprocessed_text

In [ ]:
punctuation_marks = ['!', ',', '(', ')', ':', '-', '?', '.', '..', '...']
stop_words = stopwords.words("russian")
morph = pymorphy3.MorphAnalyzer()

In [ ]:
!wget https://www.dropbox.com/s/a9r0b2yj3vqvi13/banks.csv?dl=1 -O banks.csv

--2023-09-30 07:36:37--  https://www.dropbox.com/s/a9r0b2yj3vqvi13/banks.csv?dl=1
Resolving www.dropbox.com (www.dropbox.com)... 162.125.6.18, 2620:100:601c:18::a27d:612
Connecting to www.dropbox.com (www.dropbox.com)|162.125.6.18|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: /s/dl/a9r0b2yj3vqvi13/banks.csv [following]
--2023-09-30 07:36:37--  https://www.dropbox.com/s/dl/a9r0b2yj3vqvi13/banks.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc797e4eb94195bee2c9e6eeb3a5.dl.dropboxusercontent.com/cd/0/get/CEuEldOwv4jj_CwYLKfNcxeEtEh0phMq2Uc8dfdRa3kDfPbUdTrbT7f9tY7uJTcpYj777M5d16LG_JtihmczPbpewV55yBaCesapBWbCdWWW77XNCRgukBmD6J7vHXYIFVdarrcVCU7KCWOfpNFM0fiI/file?dl=1# [following]
--2023-09-30 07:36:38--  https://uc797e4eb94195bee2c9e6eeb3a5.dl.dropboxusercontent.com/cd/0/get/CEuEldOwv4jj_CwYLKfNcxeEtEh0phMq2Uc8dfdRa3kDfPbUdTrbT7f9tY7uJTcpYj777M5d16LG_JtihmczPbpewV55yBaCesapBWbCdWWW7

In [ ]:
banks = pd.read_csv('banks.csv', sep='\t', index_col='idx');

In [ ]:
banks

Score                                               Text
idx                                                               
0      Positive  В Альфа-Банке работает замечательная девушка -...
1      Negative  Оформляя рассрочку в м. Видео в меге тёплый ст...
2      Positive  Очень порадовала оперативность работы в банке....
3      Negative  Имела неосторожность оформить потреб. кредит в...
4      Negative  Небольшая предыстория: Нашел на сайте MDM банк...
...         ...                                                ...
13994  Positive  О высокой надёжности МКБ, порядочности и добро...
13995  Positive  Обслуживаюсь в офисе на Чернореченской 42а, ка...
13996  Positive  Попала сегодня в очень неприятную ситуацию. Ре...
13997  Positive  Добрый день! Давно являюсь клиентом банка Русс...
13998  Negative  Сначала было все банально. Взял в кредит mp3 п...

[13999 rows x 2 columns]

In [ ]:
text = banks.iloc[0]['Text']

In [ ]:
preprocessed_text = preprocess(text, stop_words, punctuation_marks, morph)

In [ ]:
preprocessed_text

['альфа-банк',
 'работать',
 'замечательный',
 'девушка',
 'ильясова',
 'орный',
 'вежливый',
 'отзывчивый',
 'действительно',
 'участвовать',
 'запрос',
 'клиент',
 'приходить',
 'подряд',
 'ровно',
 'день',
 'каждый',
 'день',
 'помнить',
 'время',
 'приход',
 'помочь',
 'оформить',
 'кредит',
 'размер',
 '1млн',
 'рубль',
 'прийти',
 'партнёр',
 'передавать',
 'получить',
 'кредит',
 'покупать',
 'авто',
 'специалист',
 'ильясова',
 'орный',
 'помочь',
 'вывести',
 'весь',
 'сумма',
 'день',
 'это',
 'непросто',
 'сделка',
 'состояться',
 'всё',
 'остаться',
 'довольный',
 'знакомый',
 'конец',
 'весь',
 'добавить',
 'понять',
 'почему',
 'прислать',
 'это',
 'отделение',
 'девушка',
 'большой',
 'замечательный',
 'специалист',
 'приобретать',
 'программа',
 'здоровье',
 'вообще',
 'всё',
 'связать',
 'альфа-банк']

Векторные представления слов Navec для отзыва о банке

In [ ]:
navec['альфа-банк']

array([-3.88679430e-02,  1.24705888e-01,  4.56231058e-01, -3.73969600e-02,
       -3.47879499e-01, -3.50659519e-01,  2.28740707e-01, -4.75479364e-01,
       -3.07692774e-02,  1.81494340e-01,  1.90260470e-01,  5.54586589e-01,
        8.50039274e-02,  3.10050279e-01, -3.09065312e-01,  6.30352676e-01,
       -7.04480410e-02,  2.14478374e-01,  7.24898651e-02,  4.41185266e-01,
       -9.94821191e-02, -2.98542660e-02,  9.04064253e-02, -8.20990562e-01,
        1.69931337e-01,  1.56062335e-01,  3.05335402e-01,  1.96679264e-01,
       -2.16666460e-01, -4.00866151e-01,  2.71243483e-01, -4.08422202e-02,
       -3.37154150e-01, -8.11191261e-01, -5.41225195e-01, -1.68108210e-01,
        3.58279049e-02, -3.14308286e-01, -9.95728076e-02, -2.49294430e-01,
       -2.88453177e-02,  2.66260188e-03, -3.93901497e-01,  4.36253488e-01,
       -2.14055538e-01, -1.26410529e-01, -9.77082253e-02, -2.43003234e-01,
       -3.79067838e-01,  3.56710970e-01,  4.17160064e-01,  9.72895622e-02,
        3.08841318e-01,  

In [ ]:
navec['работать']

array([ 1.86381385e-01, -2.70801961e-01, -3.03715795e-01,  3.36435549e-02,
       -2.11302772e-01,  3.53677534e-02,  4.32576895e-01, -4.20273662e-01,
       -8.38170499e-02,  3.23135078e-01, -1.16680332e-01, -3.11935633e-01,
       -2.19287634e-01,  4.94534612e-01, -5.34101427e-01, -1.82629958e-01,
       -6.65567815e-01, -1.10345572e-01, -3.51800442e-01,  1.31303012e-01,
       -1.82882130e-01, -4.52215225e-01, -2.55357653e-01,  5.98089278e-01,
        2.18318000e-01, -3.65659371e-02, -5.24929047e-01,  8.35759103e-01,
        3.44622612e-01,  3.16270679e-01, -3.72660935e-01, -1.82024434e-01,
        1.17612135e+00, -1.84543774e-01,  2.11849391e-01,  7.58146420e-02,
       -2.70730108e-01, -2.07940474e-01,  7.30130374e-01,  2.52080988e-02,
       -1.68424785e-01,  9.99179482e-02,  7.30607212e-02, -3.97728026e-01,
       -1.06336057e-01,  8.51420909e-02,  1.17559962e-01, -3.79440695e-01,
        9.18142796e-02, -3.42777014e-01, -9.44375843e-02, -1.00128055e-01,
       -4.00310792e-02,  

Находим плотные векторные представления для всех слов в тексте

In [ ]:
text_embeddings = []
for tocken in preprocessed_text:
    embedding = navec[tocken]
    text_embeddings.append(embedding)

KeyError: ignored

**Учитываем неизвестные слова**

Вектор для неизвестного слова в Navec

In [ ]:
navec['<unk>']

array([ 2.14312136e-01,  3.70287180e-01,  1.36796311e-01, -1.86535835e-01,
       -4.91572991e-02, -1.88751370e-02, -1.32190725e-02,  3.57939489e-02,
       -3.97918969e-02,  1.94128811e-01, -8.25866908e-02,  2.47690390e-04,
        2.88263001e-02,  2.01173663e-01, -1.54272586e-01, -1.13935187e-01,
       -4.74858470e-02, -7.92295299e-03,  6.50205165e-02, -3.78477909e-02,
        6.24999329e-02,  2.49566153e-01,  1.03294946e-01, -2.11493388e-01,
       -1.73085947e-02, -2.82477215e-02, -6.87575415e-02, -9.21097770e-02,
        8.71437322e-03, -1.69095173e-01, -5.73454238e-02,  4.21022065e-02,
       -5.24346411e-01, -1.58332035e-01,  5.83604947e-02, -6.78519439e-03,
       -6.93208054e-02, -5.74708311e-03, -1.20353170e-01, -4.40001450e-02,
        4.75032702e-02, -2.33378902e-01, -1.33015737e-01,  1.27385199e-01,
       -7.16302916e-02,  1.28748834e-01,  1.13330811e-01,  1.26265138e-02,
        5.89972734e-02,  2.43283421e-01, -8.16499963e-02,  2.72306442e-01,
        1.67372063e-01, -

In [ ]:
unk = navec['<unk>']
text_embeddings = []
for tocken in preprocessed_text:
    embedding = navec.get(tocken, unk)
    text_embeddings.append(embedding)

Векторное представление для текста

In [ ]:
text_embeddings

[array([-3.88679430e-02,  1.24705888e-01,  4.56231058e-01, -3.73969600e-02,
        -3.47879499e-01, -3.50659519e-01,  2.28740707e-01, -4.75479364e-01,
        -3.07692774e-02,  1.81494340e-01,  1.90260470e-01,  5.54586589e-01,
         8.50039274e-02,  3.10050279e-01, -3.09065312e-01,  6.30352676e-01,
        -7.04480410e-02,  2.14478374e-01,  7.24898651e-02,  4.41185266e-01,
        -9.94821191e-02, -2.98542660e-02,  9.04064253e-02, -8.20990562e-01,
         1.69931337e-01,  1.56062335e-01,  3.05335402e-01,  1.96679264e-01,
        -2.16666460e-01, -4.00866151e-01,  2.71243483e-01, -4.08422202e-02,
        -3.37154150e-01, -8.11191261e-01, -5.41225195e-01, -1.68108210e-01,
         3.58279049e-02, -3.14308286e-01, -9.95728076e-02, -2.49294430e-01,
        -2.88453177e-02,  2.66260188e-03, -3.93901497e-01,  4.36253488e-01,
        -2.14055538e-01, -1.26410529e-01, -9.77082253e-02, -2.43003234e-01,
        -3.79067838e-01,  3.56710970e-01,  4.17160064e-01,  9.72895622e-02,
         3.0

Длина векторного представления текста

In [ ]:
len(text_embeddings)

69

Вектор для символа-заполнителя

In [ ]:
navec['<pad>']

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0.

## Полезные ссылки

1. [Библиотека Gensim](https://radimrehurek.com/gensim/index.html).
2. [Word2Vec в Gensim](https://radimrehurek.com/gensim/auto_examples/tutorials/run_word2vec.html).
3. [Библиотека fastText](https://fasttext.cc/).
4. [Предварительно обученные векторные представления для 157 языков](https://fasttext.cc/docs/en/crawl-vectors.html).
3. [Библиотека navec](https://natasha.github.io/navec/).
4. [Navec — компактные эмбеддинги для русского языка](https://natasha.github.io/navec/).